### PR TITLE
sign in

### PR DESCRIPTION
**Pull request by jonnyalexbh**
## Summary
- Sign In

## Trello Card
https://trello.com/c/8uIBe308/3-sign-in

## Screenshot
Captura de pantalla 2019-08-19 a la(s) 5 11 22 p  m

### CODE REVIEW

File: test/user.spec.js [hunk 1]
```diff
@@ -34,15 +36,44 @@ describe('User Creation', () => {
...
+          .send({ email: response.body.email, password: user.password })
+          .then(res => {
+            expect(res.status).toEqual(200);
+            expect(res.body).toHaveProperty('token');
+            done();
+          });
+      }));
+
+  it('should fail because the data does not exist', done =>
+    request(app)
```
[the remark this answers is not in the document]
Comment (jonnyalexbh) [reply test/user.spec.js]: I think factory girl is not the best solution here

File: app/services/user.js [hunk 2]
```diff
@@ -16,3 +17,13 @@ exports.createUser = user =>
       logger.error(`Could not create user: ${user.first_name}`);
       throw error.databaseError(err.message);
     });
+
+exports.login = user =>
+  User.findOne({ where: { email: user.email } }).then(result => {
```
[the remark this answers is not in the document]
Comment (jonnyalexbh) [reply app/services/user.js]: ready

File: app/utils.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const jwt = require('jwt-simple'),
+  config = require('../config/index'),
+  { secret_key } = config.common.jwt;
+
+exports.generateToken = user => {
+  const tokenPayload = {
+    lastName: user.lastName,
```
Comment (jersonlopez) [inline app/utils.js]: optional

File: test/user.spec.js [hunk 4]
```diff
@@ -1,48 +1,70 @@
...
-      }));
+      .send({ ...user, lastName: '' });
+    expect(res.status).toEqual(400);
+    expect(res.body.message[0].msg).toEqual('lastName is required.');
+    done();
+  });
+});
+
+describe('User Sign-In', () => {
+  it('should authenticate correctly', async done => {
```
Comment (jersonlopez) [inline test/user.spec.js]: more descriptive please!

File: test/user.spec.js [hunk 5]
```diff
@@ -1,48 +1,70 @@
...
-          .send({ ...user, email: 'jonny@wolox.co' })
-          .then(res => {
-            expect(res.status).toEqual(422);
-            expect(res.body.message).toEqual('the email entered already exists!');
-          });
-      }));
+      .send(user.dataValues);
+    expect(res.status).toEqual(422);
+    expect(res.body.message).toEqual('the email entered already exists!');
+    done();
```
Comment (jersonlopez) [inline test/user.spec.js]: unnecessary `done()`, please remove it

File: test/user.spec.js [hunk 6]
```diff
@@ -1,48 +1,70 @@
...
+
+describe('User Sign-In', () => {
+  it('should authenticate correctly', async done => {
+    const user = await userFactory.create({ password: helpers.encryptPassword('ubuntu2018') });
+    const res = await request(app)
+      .post('/users/sessions')
+      .send({ email: user.email, password: 'ubuntu2018' });
+    expect(res.status).toEqual(200);
+    expect(res.body).toHaveProperty('token');
+    done();
```
Comment (jersonlopez) [inline test/user.spec.js]: same here!

File: app/middlewares/validator.js [hunk 7]
```diff
@@ -36,4 +36,16 @@ const fields = [
     .withMessage('Password must be alphanumeric')
```
[the remark this answers is not in the document]
Comment (jonnyalexbh) [reply app/middlewares/validator.js]: implemented

File: app/helpers.js [hunk 8]
```diff
@@ -1,6 +1,9 @@
 const bcrypt = require('bcryptjs'),
-  salt_sync = require('../config').encryption;
+  config = require('../config/index'),
+  { salt_sync } = config.common.encryption;
 
 const salt = bcrypt.genSaltSync(parseInt(salt_sync));
 
-exports.encryptPassword = password => bcrypt.hashSync(password, salt);
+exports.encryptPassword = password => (password ? bcrypt.hashSync(password, salt) : null);
```
Comment (jersonlopez) [inline app/helpers.js]: the `()` are necessy?

**Pull request merged**